### PR TITLE
Generic Log Framework for SentenceTransformer and CrossEncoder

### PR DESCRIPTION
**Pull request by skewwhiff**
Added a generic logging framework for SentenceTransformer and Cross Encoder compatible with sbert 2.2. There's an existing pull request that's specific to Tensorboard logs (#1532). This is a generic log framework, that gives control to MLDevs who want more fine-grained logging. 

Sample usage for SentenceTransformer (with wandb):
```python
loss_names = ["contrastive_loss", "custom_loss"]
def log_callback_st(train_ix, training_steps, current_lr, loss_value):
    wandb.log({f"train/{loss_names[train_ix]}_loss": loss_value,
             f"train/{loss_names[train_ix]}_lr": current_lr,
            "train/steps": training_steps})
                          
model = SentenceTransformer(...)
train_dl =DataLoader(...)
n_epochs = 3
losses = [losses.ContrastiveLoss(model=model), MyCoolLossFn(model=model)]
model.fit(..., log_steps=len(train_dl)*n_epochs//1000, log_callable=log_callback_st)
```
Sample usage for CrossEncoder(with wandb):

```python
def log_callback_ce(training_steps, current_lr, loss_value): #NOTE: There are no joint loss training for CrossEncoders. 
    wandb.log({f"train/loss": loss_value,
             "train/learning_rate": current_lr,
             "train/steps": training_steps})
                          
model = CrossEncoder(...)
train_dl =DataLoader(...)
n_epochs = 3
model.fit(...., log_steps=len(train_dl)*n_epochs//1000, log_callable=log_callback_ce)
```

### CODE REVIEW

**Comment by nreimers** [comment]
Great, I was just about thinking to work on this and to provide this as comment to #1532. 

I will have a look

File: sentence_transformers/cross_encoder/CrossEncoder.py [hunk 1]
```diff
@@ -223,7 +223,10 @@ def fit(self,
 
                 training_steps += 1
                 if logger_callback is not None and log_steps > 0 and training_steps%log_steps==0:
-                    logger_callback(training_steps, scheduler.get_last_lr(), loss_value.item())
+                    try:
+                        logger_callback(training_steps, scheduler.get_last_lr(), loss_value.item())
+                    except Exception as e:
+                        logger.warning(("Logging error encountered. Ignoring..")
```
Comment (mi1ebsco) [inline sentence_transformers/cross_encoder/CrossEncoder.py:229]: Typo on line 229! (unmatched paren)

File: sentence_transformers/SentenceTransformer.py [hunk 2]
```diff
@@ -719,6 +725,12 @@ def fit(self,
                     if not skip_scheduler:
                         scheduler.step()
 
+                    if log_steps>0 and training_steps%log_steps==log_steps-1 and log_callable is not None:
+                        try:
+                            log_callable(train_idx, training_steps, scheduler.get_last_lr(), loss_value.item())
```
Comment (ofirnk) [inline sentence_transformers/SentenceTransformer.py:730]: I had added `global_step` to better track across epochs
```suggestion
                            log_callable(train_idx, training_steps, global_step, scheduler.get_last_lr(), loss_value.item())
```

**Review by ofirnk — APPROVED** [review]
Great addition :clap: 
Integrating with some ML monitoring becomes much easier after this PR is merged

**Comment by juanluisdb** [comment]
Hey! I'm very interested in this PR. Is there anything I can help with? Thanks!

**Comment by mi1ebsco** [comment]
Hi!  Are there plans to merge this PR soon?

**Comment by aaronrmm** [comment]
There were some bugs and suggestions not changed on skewwhiff's side. I forked theirs and made the changes:
https://github.com/aaronrmm/sentence-transformers/tree/master/sentence_transformers/cross_encoder
Let me know if I should make a fresh PR or what. Thanks

**Comment by abhinav-kashyap-asus** [comment]
Hi Is there any update on this framework :) I see that this is the most requested one with other users as well. 
It will be great if this works. There is no way to track the loss.

**Comment by zachschillaci27** [comment]
> Hi Is there any update on this framework :) I see that this is the most requested one with other users as well. It will be great if this works. There is no way to track the loss.

Hi @abhinav-kashyap-asus, I'm also hoping this feature will be implemented soon. I opened a more recent PR addressing the comments here but it's also still open. As a temporary solution you could work directly off of my repo fork here: https://github.com/zachschillaci27/sentence-transformers/tree/generic-logging

**Comment by atyshka** [comment]
@nreimers Any updates on getting this merged?

**Comment by ydennisy** [comment]
@nreimers please can we have this merged - it has been open for a year, with multiple authors providing follow up PRs.
I appreciate your work on this, but if you are closed for submissions please let us know and we can fork the project.
Thanks in advance!

File: sentence_transformers/SentenceTransformer.py [hunk 3]
```diff
@@ -576,6 +576,8 @@ def fit(self,
             save_best_model: bool = True,
             max_grad_norm: float = 1,
             use_amp: bool = False,
+            log_steps: int=0,
+            log_callback = Callable[[int, int, float, float], None]=None,
```
Comment (gongel) [inline sentence_transformers/SentenceTransformer.py:580]: log_callback --> log_callable

**Comment by shreyasavant** [comment]
@nreimers Would love to see updates with this next. Is there any good way to include a log framework currently?

**Comment by zachschillaci27** [comment]
> @nreimers Would love to see updates with this next. Is there any good way to include a log framework currently?

You can checkout my [branch](https://github.com/zachschillaci27/sentence-transformers/tree/generic-logging) and work from there. See the associated PR https://github.com/UKPLab/sentence-transformers/pull/1753

**Comment by mwmercury** [comment]
@nreimers Hello, are you still alive? Seriously.

**Comment by zachschillaci27** [comment]
> @mwmercury Hello,
> 
> I will be taking over the maintenance from now forward. A logging framework is frequently requested, and I will be looking into various approaches to see what suits the project best. This PR is certainly one of the considerations, as is going much further and implementing a `transformers`-style callback system.
> 
> * Tom Aarsen

Nice! @tomaarsen happy to help with some of the development here

**Comment by jdixosnd** [comment]
@zachschillaci27  Thanks for correcting the error. Your code works well. Just that we have to be mindful and provide 5 arguments in function definition instead of 4.

**Comment by joelsimonoff** [comment]
+1 On merging this ASAP. Honestly its not that hard to work around by subclassing `SentenceTransformer`, but this is a really nice feature nonetheless.

**Comment by tomaarsen** [comment]
This will be implemented via #2449, which includes a generic log framework via [TrainerCallback](https://huggingface.co/docs/transformers/main_classes/callback#transformers.TrainerCallback) instances. Some important callbacks such as Weights & Biases and Tensorboard are already implemented & automatically used if you have `wandb` and/or `tensorboard` installed, respectively.

I'll keep this open until that PR is merged.

- Tom Aarsen
